frmts/vrt/vrtdataset.cpp

```diff
@@ -1044,6 +1044,121 @@ GDALDataset *VRTDataset::OpenVRTProtocol(const char *pszSpec)
         return nullptr;
     }
 
+    // scan for subdataset/subdataset_n in tokens, close the source dataset and reopen with subdataset if found/valid
+    // make subdataset and subdataset_n mutually exclusive
+    bool bFound_subdataset = false;
+    for (int i = 0; i < aosTokens.size(); i++)
+    {
+        char *pszKey = nullptr;
+        const char *pszValue = CPLParseNameValue(aosTokens[i], &pszKey);
+
+        if (pszKey && pszValue)
+        {
+            char **papszSubdatasets = GDALGetMetadata(poSrcDS, "SUBDATASETS");
+
+            int nSubdatasets = CSLCount(papszSubdatasets);
+
+            if (EQUAL(pszKey, "subdataset"))
+            {
+                if (bFound_subdataset)
+                {
+                    CPLError(CE_Failure, CPLE_IllegalArg,
+                             "'subdataset' is mutually exclusive with option "
+                             "'subdataset_n'");
+                    poSrcDS->ReleaseRef();
+
+                    CPLFree(pszKey);
+                    return nullptr;
+                }
+                if (nSubdatasets > 0)
+                {
+                    bool bFound = false;
+                    for (int j = 0; j < nSubdatasets; j += 2)
+                    {
+                        char *pszSubdatasetSource =
+                            CPLStrdup(strstr(papszSubdatasets[j], "=") + 1);
```

Comment:
slightly unsafe to assume strstr() returns a not-null value. Should be checked

Reply:
and instead of putting that in a C char* with CPLStrdup() / CPLFree(), you could assign to a std::string for automatic memory management

Reply:
ok thanks!  I copied it from somewhere :)

Reply:
(from gdal_translate_bin)

Reply:
done